### PR TITLE
C++: flow through output iterators with user-defined operator= and operator*

### PR DESCRIPTION
**Pull request by rdmarsh2**
This PR adds general flow through `operator*` in `AddressFlow.qll` and `FlowVar.qll`, and adds iterator-specific flow through `operator=` in `DataFlowUtil.qll`. Those combine to support flow through iterators returned by the `std::back_inserter` function.

A few existing tests are affected by the `operator*` changes. I think those are mostly positive.

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll [hunk 1]
```diff
@@ -163,9 +163,9 @@ private module PartialDefinitions {
         valueToUpdate(convertedInner, this.getFullyConverted(), node) and
         innerDefinedExpr = convertedInner.getUnconverted() and
         (
-          innerDefinedExpr.(Call).getQualifier() = getAnIteratorAccess(collection)
+          innerDefinedExpr.(Call).getQualifier() = getCrementedExpr*(getAnIteratorAccess(collection))
```
Comment (MathiasVP) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/FlowVar.qll:166]: Is the performance of these transitive closures okay?

**Review by MathiasVP — COMMENTED** [review]
LGTM apart from a few small comments!

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll [hunk 2]
```diff
@@ -92,6 +103,17 @@ private predicate pointerToPointerStep(Expr pointerIn, Expr pointerOut) {
 
 private predicate lvalueToReferenceStep(Expr lvalueIn, Expr referenceOut) {
   lvalueIn.getConversion() = referenceOut.(ReferenceToExpr)
+  or
+  // C++ only
+  exists(Call c |
+    referenceOut = c and
+    c.getTarget().hasName(["operator*", "operator++"]) and
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll:110]: I can see that `operator++` is an `lvalueToReferenceStep`, assuming it returns a reference to `*this`, but I don't see how `operator*` can be thought of as being such a step. It should effectively be equivalent to `ReferenceToExpr`. Did you mean to type `operator--` instead?

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll [hunk 3]
```diff
@@ -52,6 +52,17 @@ private predicate lvalueToLvalueStep(Expr lvalueIn, Expr lvalueOut) {
...
+  or
+  // C++ only
+  exists(Call c |
+    lvalueOut = c and
+    c.getTarget().hasName(["operator*", "operator++"]) and
+    (
+      c.getQualifier() = lvalueIn or
+      not c.getTarget() instanceof MemberFunction and
+      c.getArgument(0) = lvalueIn
+    )
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/AddressFlow.qll:64]: I don't see how any of these are lvalue-to-lvalue steps.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll [hunk 4]
```diff
@@ -290,4 +290,21 @@ class BeginOrEndFunction extends MemberFunction, TaintFunction {
...
+  override predicate getsIterator(FunctionInput input, FunctionOutput output) {
+    input.isQualifierObject() and
+    output.isReturnValue()
+  }
+}
+
+class InserterIteratorFunction extends GetIteratorFunction {
+  InserterIteratorFunction() {
+    this.hasName(["front_inserter", "inserter", "back_inserter"]) and
+    this.getNamespace().hasName("std")
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:303]: Why not use `hasQualifiedName("std", [...])`?

File: cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp [hunk 5]
```diff
@@ -398,4 +398,41 @@ void test_vector_output_iterator(int b) {
...
+	sink(v12); // tainted [NOT DETECTED by IR]
+
+	std::vector<int>::iterator i13 = v13.begin();
+	*i13++ = source();
+	sink(v13); // tainted [NOT DETECTED by IR]
+
+	std::vector<int>::iterator i14 = v14.begin();
+	i14++;
+	*i14++ = source();
+	sink(v14); // tainted [NOT DETECTED by IR]
```
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp:414]: I'd like to see some prefix-`++` and some `--` too.

**Review by jbj — COMMENTED** [review]
Otherwise LGTM

File: cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll [hunk 6]
```diff
@@ -585,6 +585,23 @@ predicate simpleLocalFlowStep(Node nodeFrom, Node nodeTo) {
...
     nodeTo.asDefiningArgument() = call.getQualifier()
   )
+  or
+  // Iterators that have a user-defined `operator=`. Take flow from the RHS to
+  // the post-update node for the iterator.
+  // The built-in `=` case is handled by FlowVar, other user-defined `operator=`
+  // will be handled by interprocedural flow.
+  exists(IteratorPartialDefinitionNode postUpdate, Call opEquals |
+    postUpdate = nodeTo and
+    opEquals.getTarget().hasName("operator=") and
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/dataflow/internal/DataFlowUtil.qll:595]: Without having looked at this in a lot of detail yet - why do we need mentions of `"operator="`, `"operator*"` and `"operator++"` inside `dataflow/internal`?  Shouldn't this kind of thing be done by models?

File: cpp/ql/src/semmle/code/cpp/models/interfaces/Iterator.qll [hunk 7]
```diff
@@ -15,3 +15,7 @@ import semmle.code.cpp.models.Models
  * can be used to write to the iterator's underlying collection.
  */
 abstract class IteratorReferenceFunction extends Function { }
+
+abstract class GetIteratorFunction extends Function {
```
[the remark this answers is not in the document]
Comment (rdmarsh2) [reply cpp/ql/src/semmle/code/cpp/models/interfaces/Iterator.qll:22]: Added

**Comment by rdmarsh2** [comment]
I've rewritten most of the QL changes. It now works with just the previously existing nodes and some additional flow edges to the return value input for `operator*` and `operator=`.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll [hunk 8]
```diff
@@ -78,6 +78,20 @@ private FunctionInput getIteratorArgumentInput(Operator op, int index) {
...
+    t =
+      op
+          .getACallToThisFunction()
+          .getArgument(index)
+          .getExplicitlyConverted()
+          .getType()
+          .stripTopLevelSpecifiers()
+  |
+    result.isParameterDeref(index) // TODO: does this work with an rvalue reference?
+  )
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:92]: There's something missing here. There should be a second use of `t`.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll [hunk 9]
```diff
@@ -92,6 +106,9 @@ class IteratorPointerDereferenceOperator extends Operator, TaintFunction, Iterat
   override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
     input = iteratorInput and
     output.isReturnValue()
+    or
+    input.isReturnValueDeref() and
+    output = getIteratorArgumentOutput(this, 0)
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:111]: Can't this be simplified to `output.isParameterDeref(0)`? The charpred should have already established that we're looking at an `operator*` for an iterator.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll [hunk 10]
```diff
@@ -274,11 +294,32 @@ class IteratorArrayMemberOperator extends MemberFunction, TaintFunction, Iterato
...
+class IteratorAssignmentMemberOperator extends MemberFunction, TaintFunction {
+  IteratorAssignmentMemberOperator() {
+    this.hasName("operator=") and
+    this.getDeclaringType() instanceof Iterator and
+    not this instanceof CopyAssignmentOperator and
+    not this instanceof MoveAssignmentOperator
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameter(0) and
```
Comment (jbj) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Iterator.qll:313]: It's not `isParameterRef`? I'm seeing reference types in https://en.cppreference.com/w/cpp/iterator/back_insert_iterator/operator%3D

File: cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp [hunk 11]
```diff
@@ -398,4 +398,55 @@ void test_vector_output_iterator(int b) {
...
+		auto it = std::back_inserter(out);
+		*it++ = source();
+		sink(out); // tainted [NOT DETECTED]
+	}
+
+	{
+		std::vector<std::string> out;
+		auto it = std::back_inserter(out);
+		*++it = std::string(source_string);
+		sink(out); // tainted [NOT DETECTED]
```
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp:443]:
```suggestion
		sink(out); // tainted [NOT DETECTED by IR]
```

File: cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp [hunk 12]
```diff
@@ -398,4 +398,55 @@ void test_vector_output_iterator(int b) {
...
+	}
+
+	{
+		std::vector<int> out;
+		auto it = std::back_inserter(out);
+		*++it = source();
+		sink(out); // tainted [NOT DETECTED]
```
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp:450]:
```suggestion
		sink(out); // tainted [NOT DETECTED by IR]
```

File: cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp [hunk 13]
```diff
@@ -398,4 +398,55 @@ void test_vector_output_iterator(int b) {
...
+		auto it = out.end();
+		*it++ = std::string(source_string);
+		sink(out); // tainted [NOT DETECTED by IR]
+	}
+
+	{
+		std::vector<std::string> out;
+		auto it = std::back_inserter(out);
+		*it++ = std::string(source_string);
+		sink(out); // tainted [NOT DETECTED]
```
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp:429]:
```suggestion
		sink(out); // tainted [NOT DETECTED by IR]
```

File: cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp [hunk 14]
```diff
@@ -398,4 +398,55 @@ void test_vector_output_iterator(int b) {
...
+	}
+
+	{
+		std::vector<int> out;
+		auto it = std::back_inserter(out);
+		*it++ = source();
+		sink(out); // tainted [NOT DETECTED]
```
Comment (jbj) [inline cpp/ql/test/library-tests/dataflow/taint-tests/vector.cpp:436]:
```suggestion
		sink(out); // tainted [NOT DETECTED by IR]
```

**Review by jbj — COMMENTED** [review]
Thanks for making these changes! The code looks much better now that the library-specific details have been moved into the model classes. I have a few more review comments.

**Review by jbj — APPROVED** [review]
This PR LGTM with the latest round of changes. The test failures are caused by semantic merge conflicts with https://github.com/github/codeql/pull/4471.

**Review by jbj — APPROVED** [review]

**Comment by jbj** [comment]
I've started https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1511/. Feel free to self-merge if there are no performance or result regressions.

**Comment by jbj** [comment]
@rdmarsh2 The slowdowns on Linux and Kamailio are too big to ignore, so please investigate. They occur in queries that don't use data flow, so perhaps they're inside the models library.

**Review by jbj — APPROVED** [review]
[The re-run of CPP-Differences](https://jenkins.internal.semmle.com/job/Changes/job/CPP-Differences/1520/) doesn't have the same performance outliers, so perhaps it was just noise.

**Pull request merged**